### ISSUE TITLE
New "changed files" PR detection has false positive (history should not be considered)

### ISSUE THREAD

**Issue by mcanouil** [super-linter/super-linter#6798]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Are you using the latest Super-linter version available?

- [x] I am using the latest Super-linter version.
- [x] I can reproduce the issue running Super-linter using complete version identifier (example: vX.Y.Z), and not just with a partial one (example: vX)

- [x] I am using the super-linter/super-linter action or container image, and not the deprecated github/super-linter action or container image.

### Are you resonably sure that it's a Super-linter issue, and not an issue related to a tool that Super-linter runs?

- [x] I think that this is a Super-linter issue.

### Current Behavior

Super Linter 7.4 is scanning for files not changed by the PR on branch update.
Instead, it's now scanning for files changed in the history regardless of the changes being in the PR or not.

This is the result of:

- https://github.com/super-linter/super-linter/pull/6687

See the following repository, in particular the PR 5:

- https://github.com/mcanouil/super-linter-7.4.0-issue/pull/5

Only `file1.md` is added/changed, but because the "update" added a commit with a change in `bad-python.py` the linter consider this a PR change.

### Expected Behavior

With `VALIDATE_ALL_CODEBASE: false`, only files actually changed by the PR should be linted, not all files changed since the reference commit used by the PR.

### Super-Linter version

```markdown
7.4.0
```

### Super-linter configuration

See
- https://github.com/mcanouil/super-linter-7.4.0-issue/blob/main/.github/workflows/linter-7.4.0.yml
- https://github.com/mcanouil/super-linter-7.4.0-issue/blob/main/.github/workflows/linter-7.3.0.yml

### Relevant log output

- https://github.com/mcanouil/super-linter-7.4.0-issue/pull/4
- https://github.com/mcanouil/super-linter-7.4.0-issue/pull/5

### Steps To Reproduce

See https://github.com/mcanouil/super-linter-7.4.0-issue

1. fork
2. make a PR to add a file without any linter issues (linter is succesful)
3. modify (on `main`) `bad-python.py` or `README.md` which have both linter issues
4. trigger a branch update on your PR from step 2
5. watch the linter fails

**Comment by ferrarimarco** [comment]
Hi, thanks for the detailed report.

Can you please double check the links to the reproduction case? I get a 404.

Thanks!

**Comment by mcanouil** [comment]
Sorry, I forgot repositories are private when I create them.

**Comment by ferrarimarco** [comment]
I had a look at your examples.

The linter that is failing is jscpd.

This looks like a duplicate of #6093.

**Comment by mcanouil** [comment]
Look at the checks of https://github.com/mcanouil/super-linter-7.4.0-issue/pull/5

<img width="811" alt="Image" src="https://github.com/user-attachments/assets/33841f05-3d9b-4437-9220-eeaf4a0da9d8" />

<img width="813" alt="Image" src="https://github.com/user-attachments/assets/8dc9c30f-63c6-4596-949d-b8a379e7ae90" />

There is only one file modified by the PR

<img width="1495" alt="Image" src="https://github.com/user-attachments/assets/c161287c-c0fa-4c01-a024-489e48dae966" />

But 7.4.0 is linting the Python script after a branch update, even though the Python script is not modified by the PR. 7.3.0 only lints files actually modified by the PR not modified in the history of the PR.

<img width="1113" alt="Image" src="https://github.com/user-attachments/assets/98dbbddf-7b15-4fa5-bdd2-1e5925ea133f" />

7.4.0 means branch update are no longer possible without the action linting every single commits.

**Comment by ferrarimarco** [comment]
Can you please re-run the failed workflow, and enable debug logging? You can do that from the workflow summary page (https://github.com/mcanouil/super-linter-7.4.0-issue/actions/runs/15115562056/job/42485236462?pr=5) -> re-run workflow

Thanks!

**Comment by mcanouil** [comment]
Same results.

**Comment by ferrarimarco** [comment]
While, we wait for the debug logs, here are some observations.

- Your PR (https://github.com/mcanouil/super-linter-7.4.0-issue/pull/5) has 3 commits: c445e89, fd3aee6 (a merge commit), 44fd1a4 (another merge commit)
- The workflow triggered by c445e89 is working as expected. Both Super-linter 7.3.0 and 7.4.0 seem to be linting only one file: `/github/workspace/file1.md`.
- The workflow triggered by fd3aee6 is working as expected. The merge commit changes the workflow files. Both Super-linter 7.3.0 and 7.4.0 seem to be linting only one file: `/github/workspace/file1.md`.
- The workflow triggered by 44fd1a4 is the one that might be misbehaving. The merge commit changes the `bad-python.py` file. Super-linter 7.3.0 lints `/github/workspace/file1.md`, while 7.4.0 lints `/github/workspace/bad-python.py` and `/github/workspace/file1.md`

**Comment by mcanouil** [comment]
I did not try the various methods to update a branch, but I don't think Super Linter should lint files not really modified by the PR.

**Comment by ferrarimarco** [comment]
No need to create additional tests for now.

I think I don't see debug logs in the latest run.

The result is going to be the same, but we'd have more verbose logging to understand what's happening.

Thanks

**Comment by mcanouil** [comment]
I noticed, I must have miss-clicked the check box.
Debug logging should now be enabled.

https://github.com/mcanouil/super-linter-7.4.0-issue/actions/runs/15115562082/job/42530290252

**Comment by ferrarimarco** [comment]
I had a look at the debug log, it's helpful.

Here are the interesting bits:

```
  2025-05-20 06:30:02 [INFO]   Successfully found GITHUB_SHA: 07bd3e0dbe576041554b429effd722f93499d324
  2025-05-20 06:30:02 [DEBUG]   Successfully found the root commit: 4a996cbb786de72e68d8e49f59eabe6affd8fb96
  2025-05-20 06:30:02 [DEBUG]   This is a pull_request event
  2025-05-20 06:30:02 [DEBUG]   Updating the current GITHUB_SHA (07bd3e0dbe576041554b429effd722f93499d324) to the pull request HEAD SHA
  2025-05-20 06:30:02 [DEBUG]   Updated GITHUB_SHA: 44fd1a410e66ef73013c4d13f21b7ff63ef39c70
  2025-05-20 06:30:02 [DEBUG]   Successfully found commit count for pull_request event: 3
  2025-05-20 06:30:02 [DEBUG]   Initializing GITHUB_BEFORE_SHA because GITHUB_EVENT_COMMIT_COUNT is defined: 3
  2025-05-20 06:30:02 [DEBUG]   Initializing and validating GITHUB_BEFORE_SHA
  2025-05-20 06:30:02 [DEBUG]   Check if the 44fd1a410e66ef73013c4d13f21b7ff63ef39c70 commit is a merge commit by checking if it has more than one parent
  2025-05-20 06:30:02 [DEBUG]   44fd1a410e66ef73013c4d13f21b7ff63ef39c70 is not the initial commit. Initializing and validating GITHUB_BEFORE_SHA
  2025-05-20 06:30:02 [DEBUG]   44fd1a410e66ef73013c4d13f21b7ff63ef39c70 git commit parents count (GIT_COMMIT_PARENTS_COUNT): 3
  2025-05-20 06:30:02 [DEBUG]   Subtract 1 from GIT_COMMIT_PARENTS_COUNT to get the actual number of merge parents because the count includes the 44fd1a410e66ef73013c4d13f21b7ff63ef39c70 commit itself. GIT_COMMIT_PARENTS_COUNT: 2
  2025-05-20 06:30:02 [DEBUG]   44fd1a410e66ef73013c4d13f21b7ff63ef39c70 is a merge commit because it has more than one parent.
  2025-05-20 06:30:02 [DEBUG]   Add the suffix to GIT_BEFORE_SHA_HEAD to get the second parent of the merge commit: 44fd1a410e66ef73013c4d13f21b7ff63ef39c70^2
  2025-05-20 06:30:02 [DEBUG]   Remove one commit from GITHUB_EVENT_COMMIT_COUNT to account for the merge commit. GITHUB_EVENT_COMMIT_COUNT: 2
  2025-05-20 06:30:02 [DEBUG]   GIT_BEFORE_SHA_HEAD: 44fd1a410e66ef73013c4d13f21b7ff63ef39c70^2~2
  2025-05-20 06:30:02 [DEBUG]   Validating GITHUB_BEFORE_SHA: 1d3d3c68658d342c9f1ff68cac3aa7d42d11353c
  2025-05-20 06:30:02 [DEBUG]   Validate that the GITHUB_BEFORE_SHA reference (1d3d3c68658d342c9f1ff68cac3aa7d42d11353c) exists in this Git repository.
  2025-05-20 06:30:02 [DEBUG]   The GITHUB_BEFORE_SHA reference (1d3d3c68658d342c9f1ff68cac3aa7d42d11353c) exists in this repository
  2025-05-20 06:30:02 [DEBUG]   Successfully found GITHUB_BEFORE_SHA: 1d3d3c68658d342c9f1ff68cac3aa7d42d11353c
[...]
  2025-05-20 06:30:08 [DEBUG]   Setting GIT_DIFF_TERM to the value of GITHUB_BEFORE_SHA: 1d3d3c68658d342c9f1ff68cac3aa7d42d11353c
  2025-05-20 06:30:08 [DEBUG]   Generating Diff with:[git -C "/github/workspace" diff-tree --no-commit-id --name-only -r --root 44fd1a410e66ef73013c4d13f21b7ff63ef39c70 1d3d3c68658d342c9f1ff68cac3aa7d42d11353c | xargs -I % sh -c 'echo "/github/workspace/%"' 2>&1]
  2025-05-20 06:30:08 [DEBUG]   RAW_FILE_ARRAY contents: /github/workspace/.github/workflows/linter-7.3.0.yml /github/workspace/.github/workflows/linter-7.4.0.yml /github/workspace/bad-python.py /github/workspace/file1.md
```

So, Super-linter `v7.4.0` uses this command to generate the changed file list:

```bash
git -C \"${GITHUB_WORKSPACE}\" diff-tree --no-commit-id --name-only -r --root ${GITHUB_SHA} ${GIT_DIFF_TERM}
```

where `GIT_DIFF_TERM` is `GITHUB_BEFORE_SHA`, in this case.

Super-linter <`v7.4.0` would have used this command to generate the diff instead:

```bash
git -C \"${GITHUB_WORKSPACE}\" diff --diff-filter=d --name-only ${DEFAULT_BRANCH}...${GITHUB_SHA}
```

The <`v7.4.0` diff command excludes files that are in the PR that have the same content (as in: no difference) with the corresponding files that are already in the main branch.

I think you have a point here, this should be fixed.
